crates/ra_ssr/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+    }
+}
+
+pub struct MatchDebugInfo {
+    node: SyntaxNode,
+    /// Our search pattern parsed as the same kind of syntax node as `node`. e.g. expression, item,
+    /// etc. Will be absent if the pattern can't be parsed as that kind.
+    pattern: Result<SyntaxNode, MatchFailureReason>,
+    matched: Result<Match, MatchFailureReason>,
+}
+
+impl std::fmt::Debug for MatchDebugInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "========= PATTERN ==========\n")?;
+        match &self.pattern {
+            Ok(pattern) => {
+                write!(f, "{:#?}", pattern)?;
+            }
+            Err(err) => {
+                write!(f, "{}", err.reason)?;
+            }
+        }
+        write!(
+            f,
+            "\n============ AST ===========\n\
+            {:#?}\n============================",
+            self.node
+        )?;
+        match &self.matched {
+            Ok(_) => write!(f, "Node matched")?,
+            Err(reason) => write!(f, "Node failed to match because: {}", reason.reason)?,
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct MatchFailureReason {
+    reason: String,
+}
+
+/// Searches a crate for pattern matches and possibly replaces them with something else.
+pub struct MatchFinder<'db> {
+    /// Our source of information about the user's code.
+    sema: Semantics<'db, ra_ide_db::RootDatabase>,
+    rules: Vec<SsrRule>,
+}
+
+impl<'db> MatchFinder<'db> {
+    pub fn new(db: &'db ra_ide_db::RootDatabase) -> MatchFinder<'db> {
+        MatchFinder { sema: Semantics::new(db), rules: Vec::new() }
+    }
+
+    pub fn add_rule(&mut self, rule: SsrRule) {
+        self.rules.push(rule);
+    }
+
+    /// Adds a search pattern. For use if you intend to only call `find_matches_in_file`. If you
+    /// intend to do replacement, use `add_rule` instead.
+    pub fn add_search_pattern(&mut self, pattern: SsrPattern) {
+        self.add_rule(SsrRule { pattern, template: "()".parse().unwrap() })
+    }
+
+    pub fn edits_for_file(&self, file_id: FileId) -> Option<TextEdit> {
+        let matches = self.find_matches_in_file(file_id);
+        if matches.matches.is_empty() {
+            None
+        } else {
+            Some(replacing::matches_to_edit(&matches))
+        }
+    }
+
+    pub fn find_matches_in_file(&self, file_id: FileId) -> SsrMatches {
+        let file = self.sema.parse(file_id);
+        let code = file.syntax();
+        let mut matches = SsrMatches::default();
+        self.find_matches(code, &None, &mut matches);
+        matches
+    }
+
+    /// Finds all nodes in `file_id` whose text is exactly equal to `snippet` and attempts to match
+    /// them, while recording reasons why they don't match. This API is useful for command
+    /// line-based debugging where providing a range is difficult.
+    pub fn debug_where_text_equal(&self, file_id: FileId, snippet: &str) -> Vec<MatchDebugInfo> {
```

Comment:
I see that a bunch of public methods here are currently unused. It's probably OK, as I imagine you plan to make use of them later, but pleas don't forget to review public API afterwards, to make sure that everything that is `pub` is actually used. 

In general, in terms of code organization, the `lib.rs` here contains quite a bit of code, with mixture of private and pub stuff. Ideally (but not necessary), the `lib.rs` should be thin and list only the public API, such that it is clear at a glance what the crate can do and how it is supposed to be used. Not that we universally succeed with it in rust-analyzer, but, for example `ra_parser` and `ra_assists` manage to have at-a-glance lib.rs API, while encapsulating quite a chunk of code inside.

Reply:
I possibly will end up using that later, but you're right, I should leave it non-public for now. Making it non-public meant that a bunch of stuff was only used from the tests file, so I moved it there for now. Can move it back later if we want it public.

I also moved a number of non public things into the submodules. I didn't follow the common pattern of defining public items in submodules then re-exporting them via pub use in the root. That pattern seems reasonable, but rustc still suggests the private type rather than the public alias in error messages, so I usually avoid the pattern for now.